drivers/SmartThings/matter-sensor/src/air-quality-sensor/init.lua

```diff
@@ -164,130 +164,103 @@ local supported_profiles =
-    level_indication_support = "-all"
+  if level_name == "-co-co2-no2-ozone-ch2o-pm1-pm25-pm10-radon-tvoc" then
+    level_name = "-all"
   end
-  if level_indication_support ~= "" then
-    profile_name = profile_name .. level_indication_support .. "-level"
+  if level_name ~= "" then
+    profile_name = profile_name .. level_name .. "-level"
   end
 
-  if #co_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-co"
-  end
-  if #co2_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-co2"
-  end
-  if #no2_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-no2"
-  end
-  if #ozone_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-ozone"
-  end
-  if #formaldehyde_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-formaldehyde"
-  end
-  if #pm1_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-pm1"
-  end
-  if #pm2_5_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-pm25"
-  end
-  if #pm10_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-pm10"
-  end
-  if #radon_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-radon"
-  end
-  if #tvoc_meas_eps > 0 then
-    numeric_measurement_support = numeric_measurement_support .. "-tvoc"
-  end
   -- If all endpoints are supported, use '-all' in the profile name so that it
   -- remains under the profile name character limit
-  if numeric_measurement_support == "-co-co2-no2-ozone-formaldehyde-pm1-pm25-pm10-radon-tvoc" then
-    numeric_measurement_support = "-all"
+  if meas_name == "-co-co2-no2-ozone-ch2o-pm1-pm25-pm10-radon-tvoc" then
+    meas_name = "-all"
   end
-  if numeric_measurement_support ~= "" then
-    profile_name = profile_name .. numeric_measurement_support .. "-meas"
+  if meas_name ~= "" then
+    profile_name = profile_name .. meas_name .. "-meas"
   end
 
   if not tbl_contains(supported_profiles, profile_name) then
     device.log.warn_with({hub_logs=true}, string.format("No matching profile for device. Tried to use profile %s", profile_name))
-    if #co_meas_eps > 0 or #no2_meas_eps > 0 or #ozone_meas_eps > 0 or #formaldehyde_meas_eps > 0 or
-        #pm1_meas_eps > 0 or #pm10_meas_eps > 0 or #radon_meas_eps > 0 then
+
+    local function meas_find(sub_name)
+      return string.match(meas_name, sub_name)
+    end
+
+    -- try to best match to existing profiles
+    -- the check meas_find("co-") includes a dash to distinguish the "co" case from the "co2" case.
+    -- However, the one case where there wouldn't be a dash is when meas_name ends with "co".
+    -- To catch this edge case, we use this check- meas_name:match("co$").
+    if profile_name == meas_name:match("co$") or meas_find("co-") or meas_find("no2") or meas_find("ozone") or meas_find("ch2o") or
```

Comment:
I believe you could simplify this further by string matching this expression: `"co[^%d]?"`

This _should_ mean that we want to match any case where `co` appears and it is not (^) followed by a digit (%d), with the additional `?` at the end to account for cases where co is at the end of a string.

Lua doesn't support standard regular expressions, but it does have some functionality that you can read up on here: https://www.lua.org/manual/5.4/manual.html#6.4.1

I think distilling this to one search directly for "just co, not co2" would be more clear, along with a short comment explaining why. However, if you disagree I leave it to your discretion whichever is easier to understand.

Reply:
I don't feel super strongly about it, and it would definitely shorten the commentary needed, so let's do it :+1: thanks for the lua resource!